src/domainobjectfactories/settlement_instruction_factory.py

```diff
@@ -0,0 +1,295 @@
+import random
+from datetime import datetime, timezone, timedelta
+
+from domainobjectfactories.creatable import Creatable
+
+
+class SettlementInstructionFactory(Creatable):
+    FUNCTIONS = ['CANCEL', 'NEW']
+    LINKAGE_TYPE = ['BEFORE', 'AFTER', 'WITH', 'INFO']
+    ACCOUNT_TYPE = ['SAFE', 'CASH']
+    INSTRUCTION_TYPE = ['DVP', 'RVP', 'DELIVERY FREE', 'RECEIVABLE FREE']
+    STATUS = ['MATCHED', 'UNMATCHED']
+    message_reference_list = []
+
+    def create(self, record_count, start_id):
+        """ Create a set number of settlement instructions
+
+        Parameters
+        ----------
+        record_count : int
+            Number of settlement instructions to create
+        start_id : int
+            Starting id to use when creating message references
+
+        Returns
+        -------
+        List
+            Containing 'record_count' settlement instructions
+        """
+
+        message_reference_beginning = self.create_random_string(10)
+        self.message_reference_list = self.retrieve_column(
+            "settlement_instructions", "message_reference")
+
+        records = []
+
+        for i in range(start_id, record_count + start_id):
+            record = self.__create_record(i, message_reference_beginning)
+            records.append(record)
+            self.persist_record(
+                [record['message_reference']]
+            )
```

Comment:
Persisting message references to the database is not doing anything since we only retrieve these from the database in line 32, before anything has been persisted to start with (ie we retrieve an empty list every time). As well as persisting to database, we are appending to the instance variable `self.message_reference_list` in line 67, then selecting from this list in lines 63 and 147. The message refs in the database are never actually retrieved or used after being persisted.

Best bet is to refactor such that we still persist in line 40, but we should get rid of the `self.message_reference_list` variable completely and just retrieve a random row from the database like so:

```
def __get_linked_message(self):
    """ 50/50 chance of returning EMPTY or the message reference of
    a previously generated settlement instruction
    Returns
    -------
    String
        EMPTY or the message reference of
        a previously generated settlement instruction
    """
    linked_message = "EMPTY"
    try:
        message_reference = self.get_random_row('settlement_instructions')['message_reference']
        linked_message = random.choice([message_reference, "EMPTY"])
    except IndexError: # thrown by random.choice(self.retrieve_records(table_name)) if records retrieved are empty
        pass # use initial value "EMPTY"
    finally:
        return linked_message
```